// @wojtekmaj/react-hooks/dist/cjs/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useWindowWidth = exports.useWindowHeight = exports.useToggle = exports.useTick = exports.useSetTimeout = exports.useSetInterval = exports.useScrollTopPercent = exports.useScrollTop = exports.useScrollLeftPercent = exports.useScrollLeft = exports.useResizeObserver = exports.usePrefersReducedTransparency = exports.usePrefersReducedMotion = exports.usePrefersColorSchemeLight = exports.usePrefersColorSchemeDark = exports.usePermissionState = exports.useOnLine = exports.useMutationObserver = exports.useMatchMedia = exports.useLocalStorage = exports.useIntersectionObserver = exports.useEventListener = exports.useCurrentPosition = void 0;
var useCurrentPosition_1 = __importDefault(require("./useCurrentPosition"));
exports.useCurrentPosition = useCurrentPosition_1.default;
var useEventListener_1 = __importDefault(require("./useEventListener"));
exports.useEventListener = useEventListener_1.default;
var useIntersectionObserver_1 = __importDefault(require("./useIntersectionObserver"));
exports.useIntersectionObserver = useIntersectionObserver_1.default;
var useLocalStorage_1 = __importDefault(require("./useLocalStorage"));
exports.useLocalStorage = useLocalStorage_1.default;
var useMatchMedia_1 = __importDefault(require("./useMatchMedia"));
exports.useMatchMedia = useMatchMedia_1.default;
var useMutationObserver_1 = __importDefault(require("./useMutationObserver"));
exports.useMutationObserver = useMutationObserver_1.default;
var useOnLine_1 = __importDefault(require("./useOnLine"));
exports.useOnLine = useOnLine_1.default;
var usePermissionState_1 = __importDefault(require("./usePermissionState"));
exports.usePermissionState = usePermissionState_1.default;
var usePrefersColorSchemeDark_1 = __importDefault(require("./usePrefersColorSchemeDark"));
exports.usePrefersColorSchemeDark = usePrefersColorSchemeDark_1.default;
var usePrefersColorSchemeLight_1 = __importDefault(require("./usePrefersColorSchemeLight"));
exports.usePrefersColorSchemeLight = usePrefersColorSchemeLight_1.default;
var usePrefersReducedMotion_1 = __importDefault(require("./usePrefersReducedMotion"));
exports.usePrefersReducedMotion = usePrefersReducedMotion_1.default;
var usePrefersReducedTransparency_1 = __importDefault(require("./usePrefersReducedTransparency"));
exports.usePrefersReducedTransparency = usePrefersReducedTransparency_1.default;
var useResizeObserver_1 = __importDefault(require("./useResizeObserver"));
exports.useResizeObserver = useResizeObserver_1.default;
var useScrollLeft_1 = __importDefault(require("./useScrollLeft"));
exports.useScrollLeft = useScrollLeft_1.default;
var useScrollLeftPercent_1 = __importDefault(require("./useScrollLeftPercent"));
exports.useScrollLeftPercent = useScrollLeftPercent_1.default;
var useScrollTop_1 = __importDefault(require("./useScrollTop"));
exports.useScrollTop = useScrollTop_1.default;
var useScrollTopPercent_1 = __importDefault(require("./useScrollTopPercent"));
exports.useScrollTopPercent = useScrollTopPercent_1.default;
var useSetInterval_1 = __importDefault(require("./useSetInterval"));
exports.useSetInterval = useSetInterval_1.default;
var useSetTimeout_1 = __importDefault(require("./useSetTimeout"));
exports.useSetTimeout = useSetTimeout_1.default;
var useTick_1 = __importDefault(require("./useTick"));
exports.useTick = useTick_1.default;
var useToggle_1 = __importDefault(require("./useToggle"));
exports.useToggle = useToggle_1.default;
var useWindowHeight_1 = __importDefault(require("./useWindowHeight"));
exports.useWindowHeight = useWindowHeight_1.default;
var useWindowWidth_1 = __importDefault(require("./useWindowWidth"));
exports.useWindowWidth = useWindowWidth_1.default;
