// @llamaindex/pdf-viewer/dist/esm/constants/colors.js
export var DocumentColorEnum;
(function (DocumentColorEnum) {
    DocumentColorEnum["purple"] = "llama-purple";
    DocumentColorEnum["magenta"] = "llama-magenta";
    DocumentColorEnum["red"] = "llama-red";
    DocumentColorEnum["orange"] = "llama-orange";
    DocumentColorEnum["yellow"] = "llama-yellow";
    DocumentColorEnum["lime"] = "llama-lime";
    DocumentColorEnum["teal"] = "llama-teal";
    DocumentColorEnum["cyan"] = "llama-cyan";
    DocumentColorEnum["blue"] = "llama-blue";
    DocumentColorEnum["indigo"] = "llama-indigo";
})(DocumentColorEnum || (DocumentColorEnum = {}));
export const highlightColors = {
    [DocumentColorEnum.purple]: 'bg-llama-purple-light',
    [DocumentColorEnum.magenta]: 'bg-llama-magenta-light',
    [DocumentColorEnum.red]: 'bg-llama-red-light',
    [DocumentColorEnum.indigo]: 'bg-llama-indigo-light',
    [DocumentColorEnum.lime]: 'bg-llama-lime-light',
    [DocumentColorEnum.orange]: 'bg-llama-orange-light',
    [DocumentColorEnum.blue]: 'bg-llama-blue-light',
    [DocumentColorEnum.yellow]: 'bg-llama-yellow-light',
    [DocumentColorEnum.teal]: 'bg-llama-teal-light',
    [DocumentColorEnum.cyan]: 'bg-llama-cyan-light',
};
